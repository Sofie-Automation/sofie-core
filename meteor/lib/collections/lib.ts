import { Meteor } from 'meteor/meteor';
import { Mongo } from 'meteor/mongo';
import { TransformedCollection } from '../typings/meteor';
import { stringifyObjects, getHash } from '../lib';
import * as _ from 'underscore';
import { logger } from '../logging';

const ObserveChangeBufferTimeout = 2000;

type Timeout = number;

export function ObserveChangesForHash<Ta, Tb>(
	collection: TransformedCollection<Ta, Tb>,
	hashName: string,
	hashFields: string[],
	skipEnsureUpdatedOnStart?: boolean
) {
	const doUpdate = (id: string, obj: any) => {
		const newHash = getHash(stringifyObjects(_.pick(obj, ...hashFields)));

		if (newHash !== obj[hashName]) {
			logger.debug('Updating hash:', id, hashName + ':', newHash);
			const update = {};
			update[hashName] = newHash;
			collection.update(id, { $set: update });
		}
	};

	let observedChangesTimeouts: {
		[id: string]: Timeout;
	} = {};

	collection.find().observeChanges({
		changed: (id, changedFields) => {
			// Ignore the hash field, to stop an infinite loop
			delete changedFields[hashName];

			if (_.keys(changedFields).length > 0) {
				let data: Timeout | undefined = observedChangesTimeouts[id];
				if (data !== undefined) {
					// Already queued, so do nothing
				} else {
					// Schedule update
					observedChangesTimeouts[id] = Meteor.setTimeout(() => {
						// This looks like a race condition, but is safe as the data for the 'lost' change will still be loaded below
						delete observedChangesTimeouts[id];

						// Perform hash update
						const obj = collection.findOne(id);
						if (obj) {
							doUpdate(id, obj);
						}
					}, ObserveChangeBufferTimeout);
				}
			}
		}
	});

	if (!skipEnsureUpdatedOnStart) {
		const existing = collection.find().fetch();
		_.each(existing, (entry) => doUpdate(entry['_id'], entry));
	}
}

export function createMongoCollection<T>(
	name: string,
	options?: {
		connection?: Object | null;
		idGeneration?: string;
		transform?: Function;
	}
) {
	// Override the default mongodb methods, because the errors thrown by them doesn't contain the proper call stack

	const overrideMethod = <C>(collection: C, key: keyof C) => {
		const originalFcn: any = collection[key];

		// @ts-ignore
		collection[key] = (...args) => {
			try {
				return originalFcn.call(collection, ...args);
			} catch (e) {
				throw new Meteor.Error(
					(e && e.error) || 500,
					(e && e.reason) || e.toString() || e || 'Unknown MongoDB Error'
				);
			}
		};
	};

	const collection = new Mongo.Collection<T>(name, options);

	overrideMethod(collection, 'find');
	overrideMethod(collection, 'findOne');
	overrideMethod(collection, 'insert');
	overrideMethod(collection, 'update');
	overrideMethod(collection, 'upsert');
	overrideMethod(collection, 'remove');
	overrideMethod(collection, '_ensureIndex');
	overrideMethod(collection, '_dropIndex');

	return collection;
}
