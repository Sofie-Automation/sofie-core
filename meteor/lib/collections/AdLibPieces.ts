import { PieceGeneric } from './Pieces'
import { TransformedCollection } from '../typings/meteor'
import { registerCollection, ProtectedStringProperties, Omit } from '../lib'
import { Meteor } from 'meteor/meteor'
import { IBlueprintAdLibPiece, BaseContent } from 'tv-automation-sofie-blueprints-integration'
import { createMongoCollection } from './lib'

export interface AdLibPiece extends PieceGeneric, Omit<IBlueprintAdLibPiece, 'partId'> {
	/** The object describing the piece in detail */
	content?: BaseContent // TODO: Temporary, should be put into IBlueprintAdLibPiece

	// trigger: undefined
	disabled: false
}
export const AdLibPieces: TransformedCollection<AdLibPiece, AdLibPiece>
	= createMongoCollection<AdLibPiece>('adLibPieces')
registerCollection('AdLibPieces', AdLibPieces)
Meteor.startup(() => {
	if (Meteor.isServer) {
		AdLibPieces._ensureIndex({
			rundownId: 1,
			partId: 1,
			_rank: 1
		})
	}
})
