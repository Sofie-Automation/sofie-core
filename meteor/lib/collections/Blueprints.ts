import { registerCollection, ProtectedString } from '../lib'

import { ConfigManifestEntry, BlueprintManifestType } from '@sofie-automation/blueprints-integration'
import { createMongoCollection } from './lib'
import { OrganizationId } from './Organization'
import { registerIndex } from '../database'

/** A string, identifying a Blueprint */
export type BlueprintId = ProtectedString<'BlueprintId'>

export interface Blueprint {
	_id: BlueprintId
	organizationId: OrganizationId | null
	name: string
	/** String containing the Code for the blueprint */
	code: string
	/** Whether the blueprint has a code or not. Is equal to !!blueprint.code. */
	hasCode: boolean
	/** Timestamp, last time the blueprint was modified */
	modified: number
	/** Timestamp, when the blueprint was created */
	created: number

	blueprintId: BlueprintId
	blueprintType?: BlueprintManifestType

	studioConfigManifest?: ConfigManifestEntry[]
	showStyleConfigManifest?: ConfigManifestEntry[]

	databaseVersion: {
		showStyle: {
			[showStyleBaseId: string]: string
		}
		studio: {
			[studioId: string]: string
		}
		system: string | undefined
	}

	blueprintVersion: string
	integrationVersion: string
	TSRVersion: string
	/** Whether version checks should be disabled for this version */
	disableVersionChecks?: boolean
}

export const Blueprints = createMongoCollection<Blueprint>('blueprints')
registerCollection('Blueprints', Blueprints)

registerIndex(Blueprints, {
	organizationId: 1,
})
