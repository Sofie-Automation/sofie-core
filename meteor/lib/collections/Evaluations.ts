import { TransformedCollection } from '../typings/meteor';
import { Time, registerCollection } from '../lib';
import { Meteor } from 'meteor/meteor';
import { createMongoCollection } from './lib';
export interface Evaluation extends EvaluationBase {
	_id: string;
	userId: string;
	timestamp: Time;
}
export interface EvaluationBase {
	studioId: string;
	rundownId: string;
	answers: {
		[key: string]: string;
	};
	snapshots?: Array<string>;
}

export const Evaluations: TransformedCollection<
	Evaluation,
	Evaluation
> = createMongoCollection<Evaluation>('evaluations');
registerCollection('Evaluations', Evaluations);

Meteor.startup(() => {
	if (Meteor.isServer) {
		Evaluations._ensureIndex({
			timestamp: 1
		});
	}
});
