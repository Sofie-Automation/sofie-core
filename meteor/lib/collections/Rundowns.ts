import { Time, registerCollection, ProtectedString, ProtectedStringProperties } from '../lib'
import { Segments, Segment } from './Segments'
import { Parts, Part, DBPart } from './Parts'
import { FindOptions, MongoQuery } from '../typings/meteor'
import { StudioId } from './Studios'
import { Meteor } from 'meteor/meteor'
import { IBlueprintRundownDB } from '@sofie-automation/blueprints-integration'
import { ShowStyleVariant, ShowStyleVariants } from './ShowStyleVariants'
import { ShowStyleBase, ShowStyleBases, ShowStyleBaseId } from './ShowStyleBases'
import { RundownNote } from '../api/notes'
import { RundownPlaylists, RundownPlaylist, RundownPlaylistId } from './RundownPlaylists'
import { createMongoCollection } from './lib'
import { PeripheralDeviceId } from './PeripheralDevices'
import { OrganizationId } from './Organization'
import { registerIndex } from '../database'

export enum RundownHoldState {
	NONE = 0,
	PENDING = 1, // During STK
	ACTIVE = 2, // During full, STK is played
	COMPLETE = 3, // During full, full is played
}

export interface RundownImportVersions {
	studio: string
	showStyleBase: string
	showStyleVariant: string
	blueprint: string

	core: string
}
/** A string, identifying a Rundown */
export type RundownId = ProtectedString<'RundownId'>
/** This is a very uncomplete mock-up of the Rundown object */
export interface DBRundown
	extends ProtectedStringProperties<IBlueprintRundownDB, '_id' | 'playlistId' | 'showStyleVariantId'> {
	_id: RundownId
	/** ID of the organization that owns the rundown */
	organizationId: OrganizationId | null
	/** The id of the Studio this rundown is in */
	studioId: StudioId

	/** The ShowStyleBase this Rundown uses (its the parent of the showStyleVariant) */
	showStyleBaseId: ShowStyleBaseId
	/** The peripheral device the rundown originates from */
	peripheralDeviceId?: PeripheralDeviceId
	restoredFromSnapshotId?: RundownId
	created: Time
	modified: Time

	/** Revisions/Versions of various docs that when changed require the user to reimport the rundown */
	importVersions: RundownImportVersions

	status?: string
	// There should be something like a Owner user here somewhere?

	/** Is the rundown in an unsynced (has been unpublished from ENPS) state? */
	orphaned?: 'deleted' | 'from-snapshot' | 'manual'

	/** Last sent storyStatus to ingestDevice (MOS) */
	notifiedCurrentPlayingPartExternalId?: string

	/** Holds notes (warnings / errors) thrown by the blueprints during creation, or appended after */
	notes?: Array<RundownNote>

	/** External id of the Rundown Playlist to put this rundown in */
	playlistExternalId?: string
	/** Whether the end of the rundown marks a commercial break */
	endOfRundownIsShowBreak?: boolean
	/** Name (user-facing) of the external NCS this rundown came from */
	externalNRCSName: string
	/** The id of the Rundown Playlist this rundown is in */
	playlistId: RundownPlaylistId
	/** If the playlistId has ben set manually by a user in Sofie */
	playlistIdIsSetInSofie?: boolean
	/** Rank of the Rundown inside of its Rundown Playlist */
	_rank: number
	/** Whenever the baseline (RundownBaselineObjs, RundownBaselineAdLibItems, RundownBaselineAdLibActions) changes, this is changed too */
	baselineModifyHash?: string
}
export type Rundown = DBRundown

/**
 * Direct database accessors for the Rundown
 * These used to reside on the Rundown class
 */
export class RundownCollectionUtil {
	static getRundownPlaylist(rundown: Pick<DBRundown, 'playlistId'>): RundownPlaylist {
		if (!rundown.playlistId) throw new Meteor.Error(500, 'Rundown is not a part of a rundown playlist!')
		const pls = RundownPlaylists.findOne(rundown.playlistId)
		if (pls) {
			return pls
		} else throw new Meteor.Error(404, `Rundown Playlist "${rundown.playlistId}" not found!`)
	}
	static getShowStyleVariant(rundown: Pick<DBRundown, 'showStyleVariantId'>): ShowStyleVariant {
		const showStyleVariant = ShowStyleVariants.findOne(rundown.showStyleVariantId)
		if (!showStyleVariant)
			throw new Meteor.Error(404, `ShowStyleVariant "${rundown.showStyleVariantId}" not found!`)
		return showStyleVariant
	}
	static getShowStyleBase(rundown: Pick<DBRundown, 'showStyleBaseId'>): ShowStyleBase {
		const showStyleBase = ShowStyleBases.findOne(rundown.showStyleBaseId)
		if (!showStyleBase) throw new Meteor.Error(404, `ShowStyleBase "${rundown.showStyleBaseId}" not found!`)
		return showStyleBase
	}
	static getSegments(
		rundown: Pick<DBRundown, '_id'>,
		selector?: MongoQuery<Segment>,
		options?: FindOptions<Segment>
	): Segment[] {
		return Segments.find(
			{
				rundownId: rundown._id,
				...selector,
			},
			{
				sort: { _rank: 1 },
				...options,
			}
		).fetch()
	}
	static getParts(
		rundown: Pick<DBRundown, '_id'>,
		selector?: MongoQuery<Part>,
		options?: FindOptions<DBPart>,
		segmentsInOrder?: Array<Pick<Segment, '_id'>>
	): Part[] {
		const parts = Parts.find(
			{
				rundownId: rundown._id,
				...selector,
			},
			{
				sort: { _rank: 1 },
				...options,
			}
		).fetch()

		if (options?.sort) {
			// User explicitly sorted the parts
			return parts
		} else {
			// Default to sorting within the rundown
			return RundownPlaylist._sortPartsInner(
				parts,
				segmentsInOrder || RundownCollectionUtil.getSegments(rundown, undefined, { fields: { _id: 1 } })
			)
		}
	}
	/** Synchronous version of getSegmentsAndParts, to be used client-side */
	static getSegmentsAndPartsSync(rundown: Pick<DBRundown, '_id'>): { segments: Segment[]; parts: Part[] } {
		const segments = Segments.find(
			{
				rundownId: rundown._id,
			},
			{ sort: { _rank: 1 } }
		).fetch()

		const parts = Parts.find(
			{
				rundownId: rundown._id,
			},
			{ sort: { _rank: 1 } }
		).fetch()

		return {
			segments: segments,
			parts: RundownPlaylist._sortPartsInner(parts, segments),
		}
	}
}

export const Rundowns = createMongoCollection<Rundown>('rundowns')
registerCollection('Rundowns', Rundowns)

registerIndex(Rundowns, {
	playlistId: 1,
})
registerIndex(Rundowns, {
	playlistExternalId: 1,
})
