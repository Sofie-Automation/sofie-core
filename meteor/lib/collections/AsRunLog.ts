import { TransformedCollection } from '../typings/meteor'
import { Time, registerCollection, ProtectedString, ProtectedStringProperties } from '../lib'
import { IBlueprintAsRunLogEvent } from '@sofie-automation/blueprints-integration'
import { createMongoCollection } from './lib'
import { RundownId } from './Rundowns'
import { StudioId } from './Studios'
import { SegmentId } from './Segments'
import { PartInstanceId } from './PartInstances'
import { PieceInstanceId } from './PieceInstances'
import { registerIndex } from '../database'

export type AsRunLogEventBase = Omit<AsRunLogEvent, '_id' | 'timestamp' | 'rehersal'>

/** A string, identifying a AsRunLogEvent */
export type AsRunLogEventId = ProtectedString<'AsRunLogEventId'>

export interface AsRunLogEvent
	extends ProtectedStringProperties<
		IBlueprintAsRunLogEvent,
		'_id' | 'rundownId' | 'studioId' | 'segmentId' | 'partInstanceId' | 'pieceInstanceId'
	> {
	_id: AsRunLogEventId
	/** Timestamp of the event */
	timestamp: Time
	/** If the event was done in rehersal */
	rehersal: boolean

	rundownId: RundownId
	studioId: StudioId
	segmentId?: SegmentId
	partInstanceId?: PartInstanceId
	pieceInstanceId?: PieceInstanceId
}

export const AsRunLog: TransformedCollection<AsRunLogEvent, AsRunLogEvent> = createMongoCollection<AsRunLogEvent>(
	'asRunLog'
)
registerCollection('AsRunLog', AsRunLog)

registerIndex(AsRunLog, {
	rundownId: 1,
})
registerIndex(AsRunLog, {
	timestamp: 1,
})
