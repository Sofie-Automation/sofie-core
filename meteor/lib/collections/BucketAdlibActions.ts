import { PieceId } from './Pieces'
import { TransformedCollection } from '../typings/meteor'
import { registerCollection } from '../lib'
import { IBlueprintActionManifest } from '@sofie-automation/blueprints-integration'
import { createMongoCollection } from './lib'
import { RundownImportVersions } from './Rundowns'
import { StudioId } from './Studios'
import { ShowStyleVariantId } from './ShowStyleVariants'
import { BucketId } from './Buckets'
import { registerIndex } from '../database'
import { AdLibActionId } from './AdLibActions'

export interface BucketAdLibAction extends Omit<IBlueprintActionManifest, 'partId'> {
	_id: AdLibActionId
	bucketId: BucketId

	externalId: string

	/**
	 * If an AdLib within the Bucket doesn't match the studioId/showStyleVariantId combination
	 * the adLib will be shown as disabled
	 */
	studioId: StudioId
	showStyleVariantId: ShowStyleVariantId
	importVersions: RundownImportVersions // TODO - is this good?
}

export const BucketAdLibActions: TransformedCollection<BucketAdLibAction, BucketAdLibAction> = createMongoCollection<
	BucketAdLibAction
>('bucketAdlibActions')
registerCollection('BucketAdLibActions', BucketAdLibActions)

registerIndex(BucketAdLibActions, {
	bucketId: 1,
	studioId: 1,
})
