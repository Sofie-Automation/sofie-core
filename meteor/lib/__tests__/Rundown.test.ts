import { testInFiber } from '../../__mocks__/helpers/jest'
import { setupDefaultStudioEnvironment, DefaultEnvironment, setupDefaultRundown } from '../../__mocks__/helpers/database'
import { getResolvedSegment } from '../Rundown'
import { Rundowns } from '../collections/Rundowns'


describe('lib/Rundown', () => {
	let env: DefaultEnvironment
	let rundownId: string
	beforeAll(() => {
		env = setupDefaultStudioEnvironment()
		rundownId = setupDefaultRundown(env)
	})
	testInFiber('getResolvedSegment', () => {

		const showStyleBase = env.showStyleBase
		const rundown = Rundowns.findOne(rundownId)
		if (!rundown) throw new Error('Rundown not found')

		const segments = rundown.getSegments()
		const segment = segments[0]
		const nextSegment = segments[1]

		const resolvedSegment = getResolvedSegment(
			showStyleBase,
			rundown,
			segment,
			true
		)
		expect(resolvedSegment).toBeTruthy()
		expect(resolvedSegment.parts).toHaveLength(2)
		expect(resolvedSegment).toMatchObject({
			// segmentExtended: SegmentExtended,
			// parts: Array<PartExtended>,
			isLiveSegment: false,
			isNextSegment: false,
			currentLivePart: undefined,
			hasRemoteItems: false,
			hasGuestItems: false,
			hasAlreadyPlayed: false,
			autoNextPart: false,
			followingPart: nextSegment.getParts()[0]
		})
		expect(resolvedSegment).toMatchSnapshot()
	})
})
