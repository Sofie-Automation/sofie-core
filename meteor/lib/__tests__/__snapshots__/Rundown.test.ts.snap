// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/Rundown getResolvedSegment 1`] = `
Object {
  "autoNextPart": false,
  "currentLivePart": undefined,
  "followingPart": Part {
    "_id": "randomId9011_part1_0",
    "_rank": 10,
    "externalId": "MOCK_PART_1_0",
    "pieces": Array [],
    "renderedDuration": 0,
    "rundownId": "randomId9011",
    "segmentId": "randomId9011_segment1",
    "startsAt": 0,
    "title": "Part 1 0",
    "typeVariant": "",
    "willProbablyAutoNext": false,
  },
  "hasAlreadyPlayed": false,
  "hasGuestItems": false,
  "hasRemoteItems": false,
  "isLiveSegment": false,
  "isNextSegment": false,
  "parts": Array [
    Part {
      "_id": "randomId9011_part0_0",
      "_rank": 0,
      "duration": 20,
      "externalId": "MOCK_PART_0_0",
      "pieces": Array [
        Object {
          "_id": "randomId9011_piece000",
          "enable": Object {
            "start": 0,
          },
          "externalId": "MOCK_PIECE_000",
          "name": "Piece 000",
          "outputLayer": Object {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": Array [
              Object {
                "_id": "cam0",
                "_rank": 0,
                "followingItems": Array [],
                "name": "Camera",
                "pieces": Array [
                  [Circular],
                  Object {
                    "_id": "randomId9011_piece010",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_010",
                    "name": "Piece 010",
                    "outputLayer": [Circular],
                    "outputLayerId": "pgm",
                    "partId": "randomId9011_part0_0",
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "rundownId": "randomId9011",
                    "sourceLayer": [Circular],
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                ],
                "type": 1,
              },
              Object {
                "_id": "vt0",
                "_rank": 1,
                "followingItems": Array [],
                "name": "VT",
                "pieces": Array [
                  Object {
                    "_id": "randomId9011_piece001",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_001",
                    "name": "Piece 001",
                    "outputLayer": [Circular],
                    "outputLayerId": "pgm",
                    "partId": "randomId9011_part0_0",
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "rundownId": "randomId9011",
                    "sourceLayer": [Circular],
                    "sourceLayerId": "vt0",
                    "status": 0,
                  },
                ],
                "type": 2,
              },
            ],
            "used": true,
          },
          "outputLayerId": "pgm",
          "partId": "randomId9011_part0_0",
          "renderedDuration": null,
          "renderedInPoint": 0,
          "rundownId": "randomId9011",
          "sourceLayer": Object {
            "_id": "cam0",
            "_rank": 0,
            "followingItems": Array [],
            "name": "Camera",
            "pieces": Array [
              [Circular],
              Object {
                "_id": "randomId9011_piece010",
                "enable": Object {
                  "start": 0,
                },
                "externalId": "MOCK_PIECE_010",
                "name": "Piece 010",
                "outputLayer": Object {
                  "_id": "pgm",
                  "_rank": 0,
                  "isPGM": true,
                  "name": "PGM",
                  "sourceLayers": Array [
                    [Circular],
                    Object {
                      "_id": "vt0",
                      "_rank": 1,
                      "followingItems": Array [],
                      "name": "VT",
                      "pieces": Array [
                        Object {
                          "_id": "randomId9011_piece001",
                          "enable": Object {
                            "start": 0,
                          },
                          "externalId": "MOCK_PIECE_001",
                          "name": "Piece 001",
                          "outputLayer": [Circular],
                          "outputLayerId": "pgm",
                          "partId": "randomId9011_part0_0",
                          "renderedDuration": null,
                          "renderedInPoint": 0,
                          "rundownId": "randomId9011",
                          "sourceLayer": [Circular],
                          "sourceLayerId": "vt0",
                          "status": 0,
                        },
                      ],
                      "type": 2,
                    },
                  ],
                  "used": true,
                },
                "outputLayerId": "pgm",
                "partId": "randomId9011_part0_0",
                "renderedDuration": null,
                "renderedInPoint": 0,
                "rundownId": "randomId9011",
                "sourceLayer": [Circular],
                "sourceLayerId": "cam0",
                "status": 0,
              },
            ],
            "type": 1,
          },
          "sourceLayerId": "cam0",
          "status": 0,
        },
        Object {
          "_id": "randomId9011_piece001",
          "enable": Object {
            "start": 0,
          },
          "externalId": "MOCK_PIECE_001",
          "name": "Piece 001",
          "outputLayer": Object {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": Array [
              Object {
                "_id": "cam0",
                "_rank": 0,
                "followingItems": Array [],
                "name": "Camera",
                "pieces": Array [
                  Object {
                    "_id": "randomId9011_piece000",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_000",
                    "name": "Piece 000",
                    "outputLayer": [Circular],
                    "outputLayerId": "pgm",
                    "partId": "randomId9011_part0_0",
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "rundownId": "randomId9011",
                    "sourceLayer": [Circular],
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                  Object {
                    "_id": "randomId9011_piece010",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_010",
                    "name": "Piece 010",
                    "outputLayer": [Circular],
                    "outputLayerId": "pgm",
                    "partId": "randomId9011_part0_0",
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "rundownId": "randomId9011",
                    "sourceLayer": [Circular],
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                ],
                "type": 1,
              },
              Object {
                "_id": "vt0",
                "_rank": 1,
                "followingItems": Array [],
                "name": "VT",
                "pieces": Array [
                  [Circular],
                ],
                "type": 2,
              },
            ],
            "used": true,
          },
          "outputLayerId": "pgm",
          "partId": "randomId9011_part0_0",
          "renderedDuration": null,
          "renderedInPoint": 0,
          "rundownId": "randomId9011",
          "sourceLayer": Object {
            "_id": "vt0",
            "_rank": 1,
            "followingItems": Array [],
            "name": "VT",
            "pieces": Array [
              [Circular],
            ],
            "type": 2,
          },
          "sourceLayerId": "vt0",
          "status": 0,
        },
        Object {
          "_id": "randomId9011_piece010",
          "enable": Object {
            "start": 0,
          },
          "externalId": "MOCK_PIECE_010",
          "name": "Piece 010",
          "outputLayer": Object {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": Array [
              Object {
                "_id": "cam0",
                "_rank": 0,
                "followingItems": Array [],
                "name": "Camera",
                "pieces": Array [
                  Object {
                    "_id": "randomId9011_piece000",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_000",
                    "name": "Piece 000",
                    "outputLayer": [Circular],
                    "outputLayerId": "pgm",
                    "partId": "randomId9011_part0_0",
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "rundownId": "randomId9011",
                    "sourceLayer": [Circular],
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                  [Circular],
                ],
                "type": 1,
              },
              Object {
                "_id": "vt0",
                "_rank": 1,
                "followingItems": Array [],
                "name": "VT",
                "pieces": Array [
                  Object {
                    "_id": "randomId9011_piece001",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_001",
                    "name": "Piece 001",
                    "outputLayer": [Circular],
                    "outputLayerId": "pgm",
                    "partId": "randomId9011_part0_0",
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "rundownId": "randomId9011",
                    "sourceLayer": [Circular],
                    "sourceLayerId": "vt0",
                    "status": 0,
                  },
                ],
                "type": 2,
              },
            ],
            "used": true,
          },
          "outputLayerId": "pgm",
          "partId": "randomId9011_part0_0",
          "renderedDuration": null,
          "renderedInPoint": 0,
          "rundownId": "randomId9011",
          "sourceLayer": Object {
            "_id": "cam0",
            "_rank": 0,
            "followingItems": Array [],
            "name": "Camera",
            "pieces": Array [
              Object {
                "_id": "randomId9011_piece000",
                "enable": Object {
                  "start": 0,
                },
                "externalId": "MOCK_PIECE_000",
                "name": "Piece 000",
                "outputLayer": Object {
                  "_id": "pgm",
                  "_rank": 0,
                  "isPGM": true,
                  "name": "PGM",
                  "sourceLayers": Array [
                    [Circular],
                    Object {
                      "_id": "vt0",
                      "_rank": 1,
                      "followingItems": Array [],
                      "name": "VT",
                      "pieces": Array [
                        Object {
                          "_id": "randomId9011_piece001",
                          "enable": Object {
                            "start": 0,
                          },
                          "externalId": "MOCK_PIECE_001",
                          "name": "Piece 001",
                          "outputLayer": [Circular],
                          "outputLayerId": "pgm",
                          "partId": "randomId9011_part0_0",
                          "renderedDuration": null,
                          "renderedInPoint": 0,
                          "rundownId": "randomId9011",
                          "sourceLayer": [Circular],
                          "sourceLayerId": "vt0",
                          "status": 0,
                        },
                      ],
                      "type": 2,
                    },
                  ],
                  "used": true,
                },
                "outputLayerId": "pgm",
                "partId": "randomId9011_part0_0",
                "renderedDuration": null,
                "renderedInPoint": 0,
                "rundownId": "randomId9011",
                "sourceLayer": [Circular],
                "sourceLayerId": "cam0",
                "status": 0,
              },
              [Circular],
            ],
            "type": 1,
          },
          "sourceLayerId": "cam0",
          "status": 0,
        },
      ],
      "renderedDuration": 3000,
      "rundownId": "randomId9011",
      "segmentId": "randomId9011_segment0",
      "startsAt": 0,
      "title": "Part 0 0",
      "typeVariant": "",
      "willProbablyAutoNext": false,
    },
    Part {
      "_id": "randomId9011_part0_1",
      "_rank": 1,
      "externalId": "MOCK_PART_0_1",
      "pieces": Array [],
      "renderedDuration": 3000,
      "rundownId": "randomId9011",
      "segmentId": "randomId9011_segment0",
      "startsAt": 3000,
      "title": "Part 0 1",
      "typeVariant": "",
      "willProbablyAutoNext": false,
    },
  ],
  "segmentExtended": Object {
    "_id": "randomId9011_segment0",
    "_rank": 0,
    "externalId": "MOCK_SEGMENT_0",
    "name": "Segment 0",
    "outputLayers": Object {
      "pgm": Object {
        "_id": "pgm",
        "_rank": 0,
        "isPGM": true,
        "name": "PGM",
        "sourceLayers": Array [
          Object {
            "_id": "cam0",
            "_rank": 0,
            "followingItems": Array [],
            "name": "Camera",
            "pieces": Array [
              Object {
                "_id": "randomId9011_piece000",
                "enable": Object {
                  "start": 0,
                },
                "externalId": "MOCK_PIECE_000",
                "name": "Piece 000",
                "outputLayer": [Circular],
                "outputLayerId": "pgm",
                "partId": "randomId9011_part0_0",
                "renderedDuration": null,
                "renderedInPoint": 0,
                "rundownId": "randomId9011",
                "sourceLayer": [Circular],
                "sourceLayerId": "cam0",
                "status": 0,
              },
              Object {
                "_id": "randomId9011_piece010",
                "enable": Object {
                  "start": 0,
                },
                "externalId": "MOCK_PIECE_010",
                "name": "Piece 010",
                "outputLayer": [Circular],
                "outputLayerId": "pgm",
                "partId": "randomId9011_part0_0",
                "renderedDuration": null,
                "renderedInPoint": 0,
                "rundownId": "randomId9011",
                "sourceLayer": [Circular],
                "sourceLayerId": "cam0",
                "status": 0,
              },
            ],
            "type": 1,
          },
          Object {
            "_id": "vt0",
            "_rank": 1,
            "followingItems": Array [],
            "name": "VT",
            "pieces": Array [
              Object {
                "_id": "randomId9011_piece001",
                "enable": Object {
                  "start": 0,
                },
                "externalId": "MOCK_PIECE_001",
                "name": "Piece 001",
                "outputLayer": [Circular],
                "outputLayerId": "pgm",
                "partId": "randomId9011_part0_0",
                "renderedDuration": null,
                "renderedInPoint": 0,
                "rundownId": "randomId9011",
                "sourceLayer": [Circular],
                "sourceLayerId": "vt0",
                "status": 0,
              },
            ],
            "type": 2,
          },
        ],
        "used": true,
      },
    },
    "rundownId": "randomId9011",
    "sourceLayers": Object {
      "cam0": Object {
        "_id": "cam0",
        "_rank": 0,
        "followingItems": Array [],
        "name": "Camera",
        "pieces": Array [],
        "type": 1,
      },
      "vt0": Object {
        "_id": "vt0",
        "_rank": 1,
        "followingItems": Array [],
        "name": "VT",
        "pieces": Array [],
        "type": 2,
      },
    },
  },
}
`;
